// @testdino/playwright/dist/cli/index.mjs
#!/usr/bin/env node

// src/cli/index.ts
import { Command } from "commander";
import { readFileSync } from "fs";
import { join as join4, dirname as dirname3 } from "path";
import { fileURLToPath as fileURLToPath3 } from "url";

// src/cli/config-loader.ts
import { existsSync, statSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";
import jiti from "jiti";

// src/cli/errors.ts
var TestDinoError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "TestDinoError";
    Error.captureStackTrace(this, this.constructor);
  }
};
var TokenMissingError = class extends TestDinoError {
  constructor() {
    const message = `Token is required to run tests with TestDino

Provide token via:
  \u2022 CLI flag:    npx tdpw test --token <your-token>
  \u2022 Environment: export TESTDINO_TOKEN=<your-token>
  \u2022 Config file: Create testdino.config.ts with token

Get your token at: https://testdino.com/settings`;
    super(message);
    this.name = "TokenMissingError";
  }
};
var ConfigSyntaxError = class extends TestDinoError {
  constructor(configPath, originalError) {
    const message = `Failed to load ${configPath}

${originalError.message}

Fix the syntax error and try again.`;
    super(message);
    this.name = "ConfigSyntaxError";
  }
};
var InvalidServerUrlError = class extends TestDinoError {
  constructor(url) {
    const message = `Invalid server URL: ${url}

Server URL must be a valid HTTP or HTTPS URL.

Examples:
  \u2022 https://api.testdino.com
  \u2022 https://api-v0.testdino.com
  \u2022 https://global.testdino.com`;
    super(message);
    this.name = "InvalidServerUrlError";
  }
};

// src/cli/config-loader.ts
var CONFIG_FILENAMES = ["testdino.config.ts", "testdino.config.js"];
var ConfigLoader = class {
  cwd;
  constructor(cwd = process.cwd()) {
    this.cwd = cwd;
  }
  /**
   * Load config file from current directory or parent directories
   */
  async load() {
    const configPath = this.findConfigFile();
    if (!configPath) {
      return { config: {} };
    }
    try {
      const config = this.loadConfigFile(configPath);
      return { config, configPath };
    } catch (error) {
      throw new ConfigSyntaxError(configPath, error instanceof Error ? error : new Error(String(error)));
    }
  }
  /**
   * Find config file by searching up directory tree
   * Stops at .git directory
   */
  findConfigFile() {
    let currentDir = this.cwd;
    while (true) {
      for (const filename of CONFIG_FILENAMES) {
        const configPath = join(currentDir, filename);
        if (existsSync(configPath)) {
          return configPath;
        }
      }
      const gitDir = join(currentDir, ".git");
      if (existsSync(gitDir) && statSync(gitDir).isDirectory()) {
        for (const filename of CONFIG_FILENAMES) {
          const configPath = join(currentDir, filename);
          if (existsSync(configPath)) {
            return configPath;
          }
        }
        break;
      }
      const parentDir = dirname(currentDir);
      if (parentDir === currentDir) {
        break;
      }
      currentDir = parentDir;
    }
    return void 0;
  }
  /**
   * Load and parse config file using jiti
   */
  loadConfigFile(configPath) {
    const jitiLoader = jiti(dirname(configPath), {
      interopDefault: true,
      cache: false,
      extensions: [".ts", ".js"]
    });
    let loaded;
    try {
      const resolved = jitiLoader.esmResolve(configPath, { try: true });
      if (!resolved) {
        throw new Error(`Could not resolve config file: ${configPath}`);
      }
      const resolvedPath = typeof resolved === "string" ? resolved : fileURLToPath(resolved);
      loaded = jitiLoader(resolvedPath);
    } catch (error) {
      throw new Error(`Syntax error: ${error instanceof Error ? error.message : String(error)}`);
    }
    let config;
    if (loaded && typeof loaded === "object" && "__esModule" in loaded) {
      config = loaded.default;
    } else if (loaded && typeof loaded === "object" && "default" in loaded) {
      config = loaded.default;
    } else {
      config = loaded;
    }
    if (config === null || config === void 0) {
      return {};
    }
    if (typeof config === "function") {
      try {
        config = config();
      } catch (error) {
        throw new Error(`Error executing config function: ${error instanceof Error ? error.message : String(error)}`);
      }
      if (config instanceof Promise) {
        throw new Error("Async config functions are not supported");
      }
      if (config === null || config === void 0) {
        return {};
      }
    }
    if (config && typeof config !== "object") {
      throw new Error("Config must be an object");
    }
    return config ?? {};
  }
};

// src/cli/config-detector.ts
import { existsSync as existsSync2 } from "fs";
import { join as join2, dirname as dirname2 } from "path";
import { fileURLToPath as fileURLToPath2 } from "url";
import jiti2 from "jiti";
var PLAYWRIGHT_CONFIG_FILENAMES = ["playwright.config.ts", "playwright.config.js"];
var TESTDINO_REPORTER_NAMES = ["@testdino/playwright", "testdino-playwright", "TestdinoReporter"];
function isValidThreshold(value) {
  return typeof value === "number" && Number.isFinite(value) && value >= 0 && value <= 100;
}
var ConfigDetector = class {
  cwd;
  constructor(cwd = process.cwd()) {
    this.cwd = cwd;
  }
  /**
   * Detect TestdinoReporter in Playwright config
   */
  async detect() {
    const configPath = this.findPlaywrightConfig();
    if (!configPath) {
      return { hasReporter: false };
    }
    try {
      const config = this.loadPlaywrightConfig(configPath);
      const result = this.extractTestdinoReporter(config);
      return {
        ...result,
        configPath
      };
    } catch (error) {
      throw new ConfigSyntaxError(configPath, error instanceof Error ? error : new Error(String(error)));
    }
  }
  /**
   * Find playwright.config.[ts|js] in current directory
   */
  findPlaywrightConfig() {
    for (const filename of PLAYWRIGHT_CONFIG_FILENAMES) {
      const configPath = join2(this.cwd, filename);
      if (existsSync2(configPath)) {
        return configPath;
      }
    }
    return void 0;
  }
  /**
   * Load and parse Playwright config using jiti
   */
  loadPlaywrightConfig(configPath) {
    const jitiLoader = jiti2(dirname2(configPath), {
      interopDefault: true,
      cache: false,
      extensions: [".ts", ".js"]
    });
    let loaded;
    try {
      const resolved = jitiLoader.esmResolve(configPath, { try: true });
      if (!resolved) {
        throw new Error(`Could not resolve Playwright config: ${configPath}`);
      }
      const resolvedPath = typeof resolved === "string" ? resolved : fileURLToPath2(resolved);
      loaded = jitiLoader(resolvedPath);
    } catch (error) {
      throw new Error(`Syntax error: ${error instanceof Error ? error.message : String(error)}`);
    }
    let config;
    if (loaded && typeof loaded === "object" && "__esModule" in loaded) {
      config = loaded.default;
    } else if (loaded && typeof loaded === "object" && "default" in loaded) {
      config = loaded.default;
    } else {
      config = loaded;
    }
    if (typeof config === "function") {
      try {
        config = config();
      } catch (error) {
        throw new Error(`Error executing config function: ${error instanceof Error ? error.message : String(error)}`);
      }
    }
    if (!config || typeof config !== "object") {
      throw new Error("Playwright config must be an object");
    }
    return config;
  }
  /**
   * Extract TestdinoReporter configuration from Playwright config
   */
  extractTestdinoReporter(config) {
    const { reporter } = config;
    if (!reporter) {
      return { hasReporter: false };
    }
    if (typeof reporter === "string") {
      if (this.isTestdinoReporter(reporter)) {
        return { hasReporter: true, options: {} };
      }
      return { hasReporter: false };
    }
    if (Array.isArray(reporter) && reporter.length > 0) {
      if (typeof reporter[0] === "string") {
        const [name, options] = reporter;
        if (this.isTestdinoReporter(name)) {
          return {
            hasReporter: true,
            options: this.extractOptions(options)
          };
        }
      }
      for (const item of reporter) {
        if (typeof item === "string") {
          if (this.isTestdinoReporter(item)) {
            return { hasReporter: true, options: {} };
          }
        } else if (Array.isArray(item) && item.length > 0) {
          const [name, options] = item;
          if (this.isTestdinoReporter(name)) {
            return {
              hasReporter: true,
              options: this.extractOptions(options)
            };
          }
        }
      }
    }
    return { hasReporter: false };
  }
  /**
   * Check if reporter name matches TestdinoReporter
   */
  isTestdinoReporter(name) {
    return TESTDINO_REPORTER_NAMES.some((testdinoName) => name === testdinoName || name.includes(testdinoName));
  }
  /**
   * Extract and validate TestdinoConfig options
   */
  extractOptions(options) {
    if (!options || typeof options !== "object") {
      return {};
    }
    const config = {};
    if ("token" in options && typeof options.token === "string") {
      config.token = options.token;
    }
    if ("serverUrl" in options && typeof options.serverUrl === "string") {
      config.serverUrl = options.serverUrl;
    }
    if ("ciRunId" in options && typeof options.ciRunId === "string") {
      config.ciRunId = options.ciRunId;
    }
    if ("debug" in options && typeof options.debug === "boolean") {
      config.debug = options.debug;
    }
    if ("coverage" in options && typeof options.coverage === "object" && options.coverage !== null) {
      config.coverage = this.extractCoverageConfig(options.coverage);
    }
    return config;
  }
  /**
   * Extract and validate CoverageConfig from reporter options
   */
  extractCoverageConfig(raw) {
    const config = {
      enabled: typeof raw.enabled === "boolean" ? raw.enabled : false
    };
    if (Array.isArray(raw.include)) {
      config.include = raw.include.filter((p) => typeof p === "string");
    }
    if (Array.isArray(raw.exclude)) {
      config.exclude = raw.exclude.filter((p) => typeof p === "string");
    }
    if (typeof raw.thresholds === "object" && raw.thresholds !== null) {
      const t = raw.thresholds;
      config.thresholds = {};
      if (isValidThreshold(t.statements)) config.thresholds.statements = t.statements;
      if (isValidThreshold(t.branches)) config.thresholds.branches = t.branches;
      if (isValidThreshold(t.functions)) config.thresholds.functions = t.functions;
      if (isValidThreshold(t.lines)) config.thresholds.lines = t.lines;
    }
    return config;
  }
};

// src/cli/config-merger.ts
import { randomUUID } from "crypto";
var ConfigMerger = class _ConfigMerger {
  static DEFAULT_SERVER_URL = "https://api.testdino.com";
  /**
   * Merge configurations from all sources
   * Priority (highest to lowest):
   * 1. CLI flags
   * 2. testdino.config.[ts|js]
   * 3. playwright.config reporter options
   * 4. Environment variables
   */
  merge(sources) {
    const { env = {}, playwrightConfig = {}, testdinoConfig = {}, cliOptions = {} } = sources;
    const token = this.selectValue(cliOptions.token, testdinoConfig.token, playwrightConfig.token, env.token);
    const serverUrl = this.selectValue(cliOptions.serverUrl, testdinoConfig.serverUrl, playwrightConfig.serverUrl, env.serverUrl) || _ConfigMerger.DEFAULT_SERVER_URL;
    const ciRunId = this.selectValue(cliOptions.ciRunId, testdinoConfig.ciRunId, playwrightConfig.ciRunId) || this.generateCiRunId();
    const debug = this.selectValue(cliOptions.debug, testdinoConfig.debug, playwrightConfig.debug, env.debug) ?? false;
    const artifacts = cliOptions.artifacts !== void 0 ? cliOptions.artifacts : this.selectValue(testdinoConfig.artifacts, playwrightConfig.artifacts) ?? true;
    const coverage = this.mergeCoverageConfig(cliOptions.coverage, testdinoConfig.coverage, playwrightConfig.coverage);
    const mergedConfig = {
      token,
      serverUrl,
      ciRunId,
      debug,
      artifacts,
      ...coverage ? { coverage } : {}
    };
    this.validate(mergedConfig);
    return mergedConfig;
  }
  /**
   * Merge coverage configuration from CLI flag and config files.
   * CLI --coverage only toggles enabled; other fields come from config files.
   */
  mergeCoverageConfig(cliCoverage, testdinoCoverage, playwrightCoverage) {
    const baseConfig = testdinoCoverage ?? playwrightCoverage;
    if (!baseConfig && cliCoverage === void 0) return void 0;
    if (cliCoverage !== void 0) {
      return { ...baseConfig || { enabled: false }, enabled: cliCoverage };
    }
    return baseConfig;
  }
  /**
   * Select first non-undefined value from arguments
   */
  selectValue(...values) {
    return values.find((value) => value !== void 0 && value !== null);
  }
  /**
   * Generate a unique CI run ID
   */
  generateCiRunId() {
    return `run-${randomUUID()}`;
  }
  /**
   * Validate merged configuration
   */
  validate(config) {
    if (!config.token || typeof config.token !== "string" || config.token.trim().length === 0) {
      throw new TokenMissingError();
    }
    if (config.serverUrl) {
      if (typeof config.serverUrl !== "string" || !this.isValidUrl(config.serverUrl)) {
        throw new InvalidServerUrlError(config.serverUrl);
      }
    }
  }
  /**
   * Check if string is a valid URL
   */
  isValidUrl(urlString) {
    try {
      const url = new URL(urlString);
      return url.protocol === "http:" || url.protocol === "https:";
    } catch {
      return false;
    }
  }
  /**
   * Get environment variables as config
   */
  static getEnvConfig() {
    const config = {};
    if (process.env.TESTDINO_TOKEN) {
      config.token = process.env.TESTDINO_TOKEN;
    }
    if (process.env.TESTDINO_SERVER_URL) {
      config.serverUrl = process.env.TESTDINO_SERVER_URL;
    }
    if (process.env.TESTDINO_DEBUG) {
      config.debug = process.env.TESTDINO_DEBUG === "true" || process.env.TESTDINO_DEBUG === "1";
    }
    return config;
  }
};

// src/cli/arg-filter.ts
var TESTDINO_FLAGS = ["--token", "-t", "--ci-run-id", "--server-url", "--debug", "--no-artifacts", "--coverage"];
var FLAGS_WITH_VALUES = ["--token", "-t", "--ci-run-id", "--server-url"];
var ArgFilter = class {
  /**
   * Filter TestDino-specific arguments from the argument list
   * Removes both flags and their values
   *
   * @param args - Raw command line arguments
   * @returns Filtered arguments safe to pass to Playwright
   */
  filter(args) {
    const result = [];
    let skipNext = false;
    for (let i = 0; i < args.length; i++) {
      const arg = args[i];
      if (skipNext) {
        skipNext = false;
        continue;
      }
      if (this.isTestdinoFlag(arg)) {
        if (this.isFlagWithValue(arg) && !arg.includes("=")) {
          skipNext = true;
        }
        continue;
      }
      if (this.isTestdinoFlagWithEquals(arg)) {
        continue;
      }
      result.push(arg);
    }
    return result;
  }
  /**
   * Check if argument is a TestDino flag
   */
  isTestdinoFlag(arg) {
    const flagName = arg.split("=")[0];
    return TESTDINO_FLAGS.includes(flagName);
  }
  /**
   * Check if argument is a TestDino flag with = syntax
   */
  isTestdinoFlagWithEquals(arg) {
    if (!arg.includes("=")) {
      return false;
    }
    const flagName = arg.split("=")[0];
    return TESTDINO_FLAGS.includes(flagName);
  }
  /**
   * Check if flag takes a value
   */
  isFlagWithValue(arg) {
    const flagName = arg.split("=")[0];
    return FLAGS_WITH_VALUES.includes(flagName);
  }
  /**
   * Get list of TestDino flags (for reference/testing)
   */
  static getTestdinoFlags() {
    return [...TESTDINO_FLAGS];
  }
};

// src/cli/temp-config.ts
import { writeFileSync, unlinkSync, existsSync as existsSync3 } from "fs";
import { join as join3 } from "path";
import { tmpdir } from "os";
import { randomUUID as randomUUID2 } from "crypto";

// src/cli/logger.ts
import chalk from "chalk";

// src/utils/index.ts
function isDebugEnabled() {
  return process.env.TESTDINO_DEBUG === "true" || process.env.TESTDINO_DEBUG === "1" || process.env.DEBUG === "true";
}

// src/cli/logger.ts
var Logger = class {
  debugEnabled;
  constructor(debugEnabled = false) {
    this.debugEnabled = debugEnabled;
  }
  /**
   * Enable or disable debug logging
   */
  setDebug(enabled) {
    this.debugEnabled = enabled;
  }
  /**
   * Log error message
   */
  error(message, error) {
    console.error(chalk.red(`
\u2716 Error: ${message}`));
    if (error && this.debugEnabled) {
      console.error(chalk.dim("\nStack trace:"));
      console.error(chalk.dim(error.stack || error.message));
    }
  }
  /**
   * Log warning message
   */
  warn(message) {
    console.warn(chalk.yellow(`\u26A0\uFE0F  Warning: ${message}`));
  }
  /**
   * Log info message
   */
  info(message) {
    console.log(chalk.blue(`\u2139\uFE0F  ${message}`));
  }
  /**
   * Log success message
   */
  success(message) {
    console.log(chalk.green(`\u2713 ${message}`));
  }
  /**
   * Log debug message (only if debug is enabled)
   */
  debug(message) {
    if (this.debugEnabled) {
      console.log(chalk.dim(`[DEBUG] ${message}`));
    }
  }
  /**
   * Log a blank line
   */
  newline() {
    console.log();
  }
  /**
   * Log a section header
   */
  section(title) {
    console.log(chalk.bold.cyan(`
${title}`));
  }
  /**
   * Log a list item
   */
  listItem(text) {
    console.log(`  \u2022 ${text}`);
  }
  /**
   * Log code/command
   */
  code(text) {
    console.log(chalk.gray(`  ${text}`));
  }
  /**
   * Format error for display
   */
  formatError(error) {
    if (this.debugEnabled && error.stack) {
      return error.stack;
    }
    return error.message;
  }
  /**
   * Log TestDino CLI banner
   */
  banner(version) {
    console.log(
      chalk.cyan(`
\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557
\u2551                                                           \u2551
\u2551   ${chalk.bold("TestDino Playwright")} v${version.padEnd(36)}\u2551
\u2551   ${chalk.dim("https://testdino.com")}                                  \u2551
\u2551                                                           \u2551
\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D
`)
    );
  }
};
var logger = new Logger(isDebugEnabled());

// src/cli/temp-config.ts
var TempConfigManager = class {
  tempFiles = /* @__PURE__ */ new Set();
  cleanupHandlersRegistered = false;
  exitHandler;
  sigintHandler;
  sigtermHandler;
  uncaughtExceptionHandler;
  unhandledRejectionHandler;
  logger;
  constructor(logger3) {
    this.logger = logger3 ?? new Logger();
  }
  /**
   * Create a temporary config file with the merged configuration
   * @param config - Merged configuration to write
   * @returns Temp config info with path and config
   */
  create(config) {
    const tempPath = this.generateTempPath();
    try {
      const configJson = JSON.stringify(config, null, 2);
      writeFileSync(tempPath, configJson, "utf-8");
      this.tempFiles.add(tempPath);
      if (!this.cleanupHandlersRegistered) {
        this.registerCleanupHandlers();
        this.cleanupHandlersRegistered = true;
      }
      return {
        path: tempPath,
        config
      };
    } catch (error) {
      throw new Error(
        `Failed to create temp config file: ${tempPath}
${error instanceof Error ? error.message : String(error)}`
      );
    }
  }
  /**
   * Clean up a specific temp config file
   * @param tempPath - Path to temp file to clean up
   */
  cleanup(tempPath) {
    try {
      if (existsSync3(tempPath)) {
        unlinkSync(tempPath);
      }
      this.tempFiles.delete(tempPath);
    } catch {
      this.logger.warn(`Failed to cleanup temp file: ${tempPath}`);
    }
  }
  /**
   * Clean up all tracked temp files
   */
  cleanupAll() {
    for (const tempPath of this.tempFiles) {
      this.cleanup(tempPath);
    }
    this.tempFiles.clear();
  }
  /**
   * Remove all event handlers (for testing)
   */
  removeHandlers() {
    if (this.exitHandler) {
      process.removeListener("exit", this.exitHandler);
    }
    if (this.sigintHandler) {
      process.removeListener("SIGINT", this.sigintHandler);
    }
    if (this.sigtermHandler) {
      process.removeListener("SIGTERM", this.sigtermHandler);
    }
    if (this.uncaughtExceptionHandler) {
      process.removeListener("uncaughtException", this.uncaughtExceptionHandler);
    }
    if (this.unhandledRejectionHandler) {
      process.removeListener("unhandledRejection", this.unhandledRejectionHandler);
    }
    this.cleanupHandlersRegistered = false;
  }
  /**
   * Generate unique temp file path
   */
  generateTempPath() {
    const filename = `testdino-config-${randomUUID2()}.json`;
    return join3(tmpdir(), filename);
  }
  /**
   * Register cleanup handlers for process exit and signals
   */
  registerCleanupHandlers() {
    this.exitHandler = () => {
      this.cleanupAll();
    };
    process.on("exit", this.exitHandler);
    this.sigintHandler = () => {
      if (this.sigintHandler) {
        process.removeListener("SIGINT", this.sigintHandler);
      }
    };
    process.on("SIGINT", this.sigintHandler);
    this.sigtermHandler = () => {
      if (this.sigtermHandler) {
        process.removeListener("SIGTERM", this.sigtermHandler);
      }
    };
    process.on("SIGTERM", this.sigtermHandler);
    this.uncaughtExceptionHandler = (error) => {
      this.logger.error("Uncaught exception", error);
      this.cleanupAll();
      process.exit(1);
    };
    process.on("uncaughtException", this.uncaughtExceptionHandler);
    this.unhandledRejectionHandler = (reason) => {
      this.logger.error("Unhandled rejection", reason instanceof Error ? reason : void 0);
      this.cleanupAll();
      process.exit(1);
    };
    process.on("unhandledRejection", this.unhandledRejectionHandler);
  }
  /**
   * Get list of tracked temp files (for testing)
   */
  getTempFiles() {
    return Array.from(this.tempFiles);
  }
};

// src/cli/playwright-spawner.ts
import { execa } from "execa";
var PlaywrightSpawner = class {
  logger;
  constructor(logger3) {
    this.logger = logger3 ?? new Logger();
  }
  /**
   * Spawn Playwright test process
   * @param options - Spawn options
   * @returns Spawn result with exit code
   */
  async spawn(options) {
    const { args, tempConfigPath, config, cwd = process.cwd() } = options;
    try {
      const env = {
        ...process.env,
        TESTDINO_CLI_CONFIG_PATH: tempConfigPath,
        TESTDINO_TOKEN: config.token,
        TESTDINO_SERVER_URL: config.serverUrl,
        TESTDINO_CI_RUN_ID: config.ciRunId,
        TESTDINO_DEBUG: config.debug ? "true" : "false"
      };
      const playwrightArgs = ["playwright", "test", "--reporter", "@testdino/playwright", ...args];
      const result = await execa("npx", playwrightArgs, {
        stdio: "inherit",
        // Forward stdout/stderr in real-time
        cwd,
        env,
        reject: false
        // Don't throw on non-zero exit codes
      });
      const exitCode = result.exitCode ?? 0;
      return {
        exitCode,
        success: exitCode === 0
      };
    } catch (error) {
      return this.handleSpawnError(error);
    }
  }
  /**
   * Handle spawn errors
   */
  handleSpawnError(error) {
    const execaError = error;
    if (execaError.code === "ENOENT") {
      this.logger.error("Failed to spawn Playwright");
      this.logger.newline();
      this.logger.info("Playwright is not installed or npx is not available.");
      this.logger.newline();
      this.logger.section("To install Playwright:");
      this.logger.code("npm install -D @playwright/test");
      this.logger.code("npx playwright install");
      return {
        exitCode: 1,
        success: false
      };
    }
    if (execaError.code === "EACCES") {
      this.logger.error("Permission denied when trying to spawn Playwright");
      this.logger.newline();
      this.logger.info("Please check file permissions and try again.");
      return {
        exitCode: 1,
        success: false
      };
    }
    this.logger.error("Failed to spawn Playwright");
    this.logger.newline();
    this.logger.info(`Error: ${execaError.message || String(error)}`);
    return {
      exitCode: 1,
      success: false
    };
  }
};

// src/cli/commands/test.ts
var TestCommand = class {
  configLoader;
  configDetector;
  configMerger;
  argFilter;
  tempConfigManager;
  playwrightSpawner;
  constructor(configLoader, configDetector, configMerger, argFilter, tempConfigManager, playwrightSpawner, logger3) {
    this.configLoader = configLoader || new ConfigLoader();
    this.configDetector = configDetector || new ConfigDetector();
    this.configMerger = configMerger || new ConfigMerger();
    this.argFilter = argFilter || new ArgFilter();
    this.tempConfigManager = tempConfigManager || new TempConfigManager(logger3);
    this.playwrightSpawner = playwrightSpawner || new PlaywrightSpawner(logger3);
  }
  /**
   * Execute the test command
   * @param options - CLI options from commander
   * @param args - Remaining arguments to pass to Playwright
   * @returns Spawn result with exit code
   */
  async execute(options, args) {
    let tempConfigPath;
    try {
      const testdinoConfigResult = await this.configLoader.load();
      const playwrightConfigResult = await this.configDetector.detect();
      const envConfig = ConfigMerger.getEnvConfig();
      const mergedConfig = this.configMerger.merge({
        env: envConfig,
        playwrightConfig: playwrightConfigResult.options,
        testdinoConfig: testdinoConfigResult.config,
        cliOptions: options
      });
      const tempConfigInfo = this.tempConfigManager.create(mergedConfig);
      tempConfigPath = tempConfigInfo.path;
      const filteredArgs = this.argFilter.filter(args);
      const result = await this.playwrightSpawner.spawn({
        args: filteredArgs,
        tempConfigPath,
        config: mergedConfig
      });
      return result;
    } finally {
      if (tempConfigPath) {
        this.tempConfigManager.cleanup(tempConfigPath);
      }
    }
  }
};

// src/cli/index.ts
var __filename = fileURLToPath3(import.meta.url);
var __dirname = dirname3(__filename);
function getVersion() {
  try {
    const packagePath = join4(__dirname, "../../package.json");
    const packageJson = JSON.parse(readFileSync(packagePath, "utf-8"));
    return packageJson.version;
  } catch {
    return "0.0.0";
  }
}
var logger2 = new Logger(isDebugEnabled());
function buildProgram() {
  const program = new Command().name("tdpw").description("Run Playwright tests with TestDino reporting").version(getVersion(), "-v, --version", "Output the current version").helpOption("-h, --help", "Display help for command");
  program.command("test").description("Run Playwright tests with TestDino reporter").option("-t, --token <token>", "TestDino authentication token").option("--ci-run-id <id>", "CI run ID for grouping test runs").option("--server-url <url>", "TestDino server URL").option("--debug", "Enable debug logging").option("--no-artifacts", "Disable artifact uploads (screenshots, videos, traces)").option("--coverage", "Enable code coverage collection").allowUnknownOption().allowExcessArguments().action(async (options, command) => {
    if (options.debug) {
      logger2.setDebug(true);
    }
    try {
      const args = command.args || [];
      const testCommand = new TestCommand();
      const result = await testCommand.execute(options, args);
      process.exit(result.exitCode);
    } catch (error) {
      if (error instanceof TestDinoError) {
        logger2.error(error.message);
      } else {
        logger2.error(
          error instanceof Error ? error.message : String(error),
          error instanceof Error ? error : void 0
        );
      }
      process.exit(1);
    }
  });
  return program;
}
async function main() {
  try {
    logger2.banner(getVersion());
    const program = buildProgram();
    await program.parseAsync(process.argv);
  } catch (error) {
    if (error instanceof TestDinoError) {
      logger2.error(error.message);
    } else {
      logger2.error(error instanceof Error ? error.message : String(error), error instanceof Error ? error : void 0);
    }
    process.exit(1);
  }
}
main().catch((error) => {
  logger2.error("Unexpected error", error instanceof Error ? error : void 0);
  process.exit(1);
});
